src/main/java/org/squiddev/cobalt/lib/Utf8Lib.java

```diff
@@ -149,26 +154,21 @@ private static int posRelative(int pos, int len) {
 	private static boolean isCont(byte[] s, int idx) {
 		return idx < s.length && (s[idx] & 0xC0) == 0x80;
 	}
-
 	private static boolean isCont(LuaString s, int idx) {
 		return isCont(s.bytes, idx);
 	}
 
+
 	/* An iterator for use in implementing utf8.codes. We store the bytes in the closure instead of in the iterator's
 	*  invariant state in the hopes that this is the tiniest bit faster.
 	*/
 	private static class Utf8CodesIter extends VarArgFunction {
-		byte[] s;
-
-		protected Utf8CodesIter(byte[] bytes) {
-			s = bytes;
-		}
-
 		@Override
 		public Varargs invoke(LuaState state, Varargs args) throws LuaError, UnwindThrowable {
-			// Arg 1: invariant state (nil)
+			// Arg 1: invariant state (the string)
 			// Arg 2: byte offset + 1
 			// Returns: byte offset + 1, code point
+			byte[] s = args.arg(1).checkLuaString().bytes;
```

Comment:
Oh goodness, I should have noticed this earlier sorry! This isn't entirely correct: `LuaString`s are a range over that, so `s.luaByte(i) != s.bytes[i]` (and `s.length != s.bytes.length`).

The sanest thing to do here might be to replace `byte[]` with `LuaString` everywhere, and then use `.substring` for `codepoints` - we can optimise it later.

Sorry, should have noticed this all before 🤦‍♂.

Reply:
Alright, I'm on it :+1: